sstables/sstables.cc

```diff
@@ -3311,18 +3319,26 @@ delete_atomically(std::vector<shared_sstable> ssts) {
             // Create temporary pending_delete log file.
             auto f = open_file_dma(tmp_pending_delete_log, oflags).get0();
             // Write all toc names into the log file.
-            file_output_stream_options options;
-            options.buffer_size = 4096;
-            auto w = file_writer(f, options, tmp_pending_delete_log);
-
-            for (const auto& sst : ssts) {
-                auto toc = sst->component_basename(component_type::TOC);
-                w.write(toc.c_str(), toc.size());
-                w.write("\n", 1);
-            }
+            try {
+                file_output_stream_options options;
+                options.buffer_size = 4096;
+                auto w = file_writer(f, std::move(options), tmp_pending_delete_log);
+
+                for (const auto& sst : ssts) {
+                    auto toc = sst->component_basename(component_type::TOC);
+                    w.write(toc.c_str(), toc.size());
+                    w.write("\n", 1);
+                }
 
-            w.flush();
-            w.close();
+                w.flush();
+                w.close();
+            } catch (...) {
+                f.close().handle_exception([tmp_pending_delete_log] (auto ep) {
```

Comment:
After you construct the file_writer, it is the owner of the file, so it is responsible for closing it. We should not close "f" under its fit, that may result in undefined behavior (like writing to the wrong file).

Reply:
You're right. We need to do that only if constructing the file_writer failed.